### PR TITLE
Wasd update

### PR DESCRIPTION
**Pull request by greengiant83**
**Description:**
Addresses issue: https://github.com/aframevr/aframe/issues/3351 
This allows wasd-controls to be placed in a parent entity around camera and motion controllers and have the motion controllers move with the player

```
<a-entity id="player" wasd-controls="">
	<a-camera wasd-controls-enabled="false"></a-camera>
	<a-entity id="leftHand" teleport-controls="cameraRig: #player" laser-controls="hand: left" raycaster="far:100; objects:.button"></a-entity>
	<a-entity id="rightHand" teleport-controls="cameraRig: #player" laser-controls="hand: right" raycaster="far:100; objects:.button"></a-entity>
</a-entity>
```

**Changes proposed:**
Adds logic to use an independent element for orientation when calculating movement vector. 
cameraRig can be specified explicitly. Falling back from that the control will search child elements for a camera element to use. Falling back from there, the existing behavior is used

### CODE REVIEW

**Comment by donmccurdy** [comment]
Thanks for getting this going! Could you remove `package-lock.json` from the PR? Curious how it got in, as it's on our `.gitignore` list.

**Comment by greengiant83** [comment]
I am still working on my git kung fu, but I think I got the package-lock.json pulled out of the pull request now.

**Comment by greengiant83** [comment]
Does the pull request look kosher now?  Just want to make sure you aren't waiting for me to fix anything.

**Comment by dmarcos** [comment]
@greengiant83 not yet 🙂 I made several comments requesting some changes. Thanks for staying on top of it

**Comment by greengiant83** [comment]
Damn. Where can I see the comments? I don't see them in this thread.

**Comment by dmarcos** [comment]
@greengiant83 https://github.com/aframevr/aframe/pull/3394/files

**Comment by greengiant83** [comment]
I must be blind, but I dont see anything over there except the file changes.

File: src/components/wasd-controls.js [hunk 1]
```diff
@@ -86,6 +88,30 @@ module.exports.Component = registerComponent('wasd-controls', {
...
+
+    // If we still haven't found a camera, search through this elements children
+    var childCamera = this.el.querySelector('[camera]');
+    if (childCamera) {
+      this.orientationEl = childCamera;
+      return;
+    }
+
+    // No cameras found, just use this element
+    this.orientationEl = this.el;
```
[the remark this answers is not in the document]
Comment (greengiant83) [reply src/components/wasd-controls.js]: I updated it to lose the cameraRig semantics in favor orientationEl.

**Comment by donmccurdy** [comment]
We may want to enforce that `orientationEl` be a descendant (or direct child?) of the controlled element. If we don't, there are some challenges and footguns:

```html
<a-scene>
  <a-entity rotation="0 90 0">
    <a-entity id="ctrl" rotation="0 -90 0"/>
  </a-entity>
  <a-entity id="sceneWrapper" rotation="0 45 0">
    <a-entity id="cameraRig" wasd-controls="orientationEl: #ctrl" rotation="0 -45 0">
      <a-entity camera/>
    </a-entity>
  </a-entity>
</a-scene>
```

Now obviously that's a _bad idea_ I'm showing above, but:

1. If we call `getAttribute('rotation')` on the `#ctrl` element, we're only getting its local transform.
2. If we call `setAttribute('rotation', rotation)` on the `#cameraRig` element, we're updating its local transform relative to a _different_ element's world transform. 
3. We could fix all of this in code by using `el.object3D.getWorldRotation()`, but there's a performance cost to calling `updateMatrixWorld()` that I wouldn't want to pay on every frame.

tl;dr — as nice as it would be to have a general solution that isn't tied to this "camera rig" concept, we get into edge cases pretty quickly when using this feature for anything other than a camera rig. Better to limit our complexity for now, I think.

@dmarcos preference? My thought is to log warnings if this is used outside of a parent/child setup. Then, at most, we have to multiply `orientationEl`'s rotation by the `cameraRig`'s rather than going arbitrarily through the tree.

**Comment by machenmusik** [comment]
You can tell whether you need to do (3) once and trivially, right?  
So it would know when using world rotation was needed?

**Comment by donmccurdy** [comment]
I guess that depends... do we want a faster path than `getWorldRotation()` if the camera is a child of the rig? What if the camera is a descendant of the rig at depth N > 1? Assume we want for the rig's movement in world space to be in the direction of the camera's heading in world space.

IMO, we should be opinionated. This gets complicated and slow for anything beyond a simple parent/child camera rig, so for now I would restrict what we attempt to support. Even with the basic rig we still have to handle the case where the rig is rotated independently, which isn't in this PR yet.

**Comment by machenmusik** [comment]
thanks @donmccurdy -- if you think it will be too slow beyond direct descendant, then +1 from me for warnings or errors when the assumption is violated - but please not every frame :-)

**Comment by greengiant83** [comment]
What if we use object3D's worldToLocal and localToWorld functions (which do not force updates to the matrices) to transform the velocity vector between coordinate spaces? This should allow the orientationEl to be at any position in the scene graph without any penalty for recursing the graph.

**Comment by dmarcos** [comment]
I'm fine adding a warning for now. We can polish the API for 0.9.0. I have a couple of ideas to create an `<a-entity player>` or `<a-entity avatar>` component to simplify everything.

**Comment by donmccurdy** [comment]
FYI, I have updated aframe-extras (master branch) to do pretty much what we're trying to add to `wasd-controls` here. [Docs](https://github.com/donmccurdy/aframe-extras/tree/master/src/controls). It might be nice to converge on similar behavior. The syntax i've used is:

```html
<a-entity id="rig"
          movement-controls
          position="25 0 25">
  <a-entity camera
            position="0 1.6 0"
            look-controls="pointerLockEnabled: true"></a-entity>
</a-entity>
```

Where `movement-controls="camera: [camera]"` is the default binding. I'm trying to be pretty narrow about the camera rig use case, rather than a general "steer any object with any other object" type of thing. Haven't implemented proper handling of a rotated rig yet, going to leave that for future work.

**Comment by donmccurdy** [comment]
Remaining issues here: 

- [ ] Are we OK with the `orientationEl` naming? I slightly prefer either `heading` or `camera` but realize I'm bikeshedding now. Also, we haven't used the `*El` suffix elsewhere that I know of, e.g. raycaster uses `objects` rather than `objectEls`. 
- [ ] Need some basic handling for a rotated rig. If no one gets to it first I'll make the change in aframe-extras and then share a suggested implementation here.

**Comment by ngokevin** [comment]
Is it necessary because if you either using motion controllers, or you are using WASD, not both at once, so it's not needed to move the whole camera rig with WASD. WASD moving the camera seems sufficient. Other use cases?

**Comment by donmccurdy** [comment]
See discussion in https://github.com/aframevr/aframe/issues/3351, but yeah, technically this may not be critical... If I move around with WASD before connecting an HMD you'll be teleported back to the origin, so that's a bit unpolished, but not a huge issue. 

FWIW I've thrown out the rotation parts of `universal-controls` to just focus on locomotion, renamed to `movement-controls`, so it can be used alongside `look-controls` now. Depends explicitly on a camera rig setup. I don't know if that should influence our recommendations here. I'd like to steer people toward using a camera rig, and it makes that easier if `wasd-controls` exemplifies it.

But since nothing is technically broken now, maybe we should take this off the 0.8.x milestone and leave it to decide for 0.9.0?

**Comment by ngokevin** [comment]
I think we can pass on this on grounds of benefit vs. complexity, amount of uncertainty, and inactivity. There's not much need to move the controllers alongside a camera with WASD. And I kind of like that WASD movement doesn't affect the VR position because they are separate paradigms. I'd agree with Don's comment below:

> IMO, we should be opinionated. This gets complicated and slow for anything beyond a simple parent/child camera rig, so for now I would restrict what we attempt to support. Even with the basic rig we still have to handle the case where the rig is rotated independently, which isn't in this PR yet.

Thanks for the work!
